Publish ClusterIngress status

Miciah opened the pull request:
Update the status conditions of ClusterIngress objects to indicate synchronization status and availability.

Watch Services in cluster-ingress-operator's namespace in order to reflect their status and any associated load-balancer ingresses to the corresponding ClusterIngress objects.

* `cmd/cluster-ingress-operator/main.go`: Watch services and endpoints in the operator's namespace.
* `manifests/00-cluster-role.yaml`: Allow operator to get services and endpoints.
* `pkg/apis/ingress/v1alpha1/status.go`: Add new functions for setting status on ClusterIngress objects: `setStatusCondition`, `SetStatusSyncCondition`, `SetStatusAvailableCondition`, `clusterIngressStatusEqual`, `UpdateStatus`.
* `pkg/apis/ingress/v1alpha1/types.go`: Add fields to `ClusterIngressStatus` for operator and load balancer status.
* `pkg/apis/ingress/v1alpha1/status_test.go`: Add tests.
* `pkg/stub/handler.go`: Use `SetStatusSyncCondition`, `SetStatusAvailableCondition`, and `UpdateStatus` in `syncIngressUpdate`.
Add new functions: `getClusterIngress`, `syncServiceUpdate`, `syncEndpointsUpdate`, and `syncServiceOrEndpointsUpdate`.
Use these new functions in `Handle` to handle Service and Endpoints events and to update the corresponding ClusterIngress status according to the Service and Endpoints status.

---

This PR depends on #47.

Miciah commented:
Rebased.

imcsk8 commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	err := sdk.Update(ci)
+	for tries := 3; tries > 0; tries-- {
```
Why is the loop needed?
the err variable is updated outside the loop or am i missing something?

Miciah replied:
Whoops! `err := sdk.Update(ci)` needs to be inside the loop.  Thanks!

Miciah replied:
I deleted the loop.  Now `UpdateStatus` returns an error value, and callers return any error from `UpdateStatus` to the handler so that the event gets reprocessed.

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,202 @@
+	ci.SetStatusCondition(condition)
+}
+
+// clusterIngressStatusEqual returns true iff the status conditions (ignoring
```
It's short for "if and only if".  I see it used a lot in included packages, but I can expand it.

Miciah replied:
Fixed.

Miciah commented:
Latest push deletes the retry logic in `UpdateStatus`; instead, `UpdateStatus` returns an error value, and its callers return any error from `UpdateStatus` to the handler so that the operator will retry processing the event.

imcsk8 commented:
/lgtm

pravisankar commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	numIngresses := len(ci.Status.LoadBalancer.Ingress)
+	if numIngresses != len(oldCi.Status.LoadBalancer.Ingress) {
```
Perform load balancer checks only when ci.Spec.HighAvailability.Type == ingressv1alpha1.CloudClusterIngressHA ?
When ci.Spec.HighAvailability.Type  != CloudClusterIngressHA , need to clear old Status.LoadBalancer field.

Miciah replied:
As you point out in another comment, the logic in `syncServiceUpdate` needs to be fixed for `ci.Spec.HighAvailability.Type != CloudClusterIngressHA`, but once that is fixed, I think the equality check should still always compare `.Status.LoadBalancer.Ingress`.  If we ever support changing from `CloudClusterIngressHA` to some other value, we want the equality check to return false so that we update with the status with the empty list of ingresses.  Am I missing anything?

pravisankar commented on pkg/stub/handler.go:
```diff
@@ -162,3 +214,98 @@ func getDaemonSetOwnerRef(ds *appsv1.DaemonSet) (metav1.OwnerReference, error) {
+		return nil
+	}
+
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
```
Check valid only in case of ci.Spec.HighAvailability.Type == ingressv1alpha1.CloudClusterIngressHA

Miciah replied:
Good point.  Thanks!

pravisankar commented on pkg/stub/handler.go:
```diff
@@ -162,3 +214,98 @@ func getDaemonSetOwnerRef(ds *appsv1.DaemonSet) (metav1.OwnerReference, error) {
+	}
+
+	ci.SetStatusAvailableCondition(true, "")
+	ci.Status.LoadBalancer.Ingress = service.Status.LoadBalancer.Ingress
```
Set this field only when ci.Spec.HighAvailability.Type == ingressv1alpha1.CloudClusterIngressHA ?

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,190 @@
+	if numConditions != len(oldCi.Status.Conditions) {
+		return false
+	}
+	for i := 0; i < numConditions; i++ {
```
Fixed.  Thanks!

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,190 @@
+	if numIngresses != len(oldCi.Status.LoadBalancer.Ingress) {
+		return false
+	}
+	for i := 0; i < numIngresses; i++ {
```
Fixed.  Thanks!

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,190 @@
+	operatorName, err := k8sutil.GetOperatorName()
+	if err != nil {
+		logrus.Fatalf("Failed to get operator name: %v", err)
+	}
```
I added the fields to `Handler`.  Since this means `CreateOrUpdateOperatorStatus` needs the Handler, and since it didn't really need the ClusterIngress, I changed its receiver to `h *Handler`, moved it to `handler.go`, and changed it to `createOrUpdateOperatorStatus`.

Miciah commented on pkg/stub/handler.go:
```diff
@@ -38,14 +42,57 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+
+		return nil
+	case *corev1.Service:
+		isDeleted := o.GetDeletionTimestamp() != nil
```
My understanding is that `event.Deleted` will go away that and `o.GetDeletionTimestamp()` will be the source of truth.

Miciah commented:
Latest push makes the following changes per Ravi's comments:

• Move `ClusterIngressStatusEqual` from `status.go` to `handler.go`, rename it to `clusterIngressStatusEqual`, and change its receiver from `ci *ClusterIngress` to `h *Handler`
• Add `operatorNamespace` and `operatorName` to `Handler` and initialize them in `NewHandler`.
• Use `h.operatorNamespace` and `h.operatorName` in `clusterIngressStatusEqual`.
• Change the equality check in `clusterIngressStatusEqual` to be agnostic to the ordering of conditions and ingresses.
• Change `syncServiceUpdate` to set the ClusterIngress's "Available" condition to "True" as soon as endpoints are ready without checking ingresses if the ClusterIngress's type is not `CloudClusterIngressHA`

pravisankar commented on pkg/stub/handler.go:
```diff
@@ -162,3 +230,149 @@ func getDaemonSetOwnerRef(ds *appsv1.DaemonSet) (metav1.OwnerReference, error) {
+
+	ciName := service.Name[7:]
+
+	ciNamespace, err := k8sutil.GetWatchNamespace()
```
Use h.operatorNamespace instead?

Miciah replied:
Fixed.  Thanks!

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,137 @@
+// objects are equal.
+func clusterIngressStatusEqual(oldCi, ci *ClusterIngress) bool {
+	numConditions := len(ci.Status.Conditions)
+	if numConditions != len(oldCi.Status.Conditions) {
```
Fixed.

Miciah commented on pkg/apis/ingress/v1alpha1/status.go:
```diff
@@ -0,0 +1,137 @@
+		}
+	}
+
+	numIngresses := len(ci.Status.LoadBalancer.Ingress)
```
Fixed.

Miciah commented:
Latest push makes the following changes per Ravi's comments:

• Delete `numConditions` and `numIngresses` in `clusterIngressStatusEqual`.
• Use `h.operatorNamespace` in `syncServiceUpdate`.

I made the following additional changes:

• Added some unit tests in `status_test.go`.
• Renamed `createOrUpdateOperatorStatus` to `syncOperatorStatus`.

After some further consideration, I think we should discuss during the next planning call how the operator should handle the "Available" condition when `ci.Spec.HighAvailability.Type` is not `CloudClusterIngressHA`.

Miciah commented:
Latest push removes the OperatorStatus code; this PR now depends on #47 to import github.com/openshift/cluster-version-operator, which has some required types.  This push also adds a watch on endpoints and event handling for the same in order to resolve a race condition that would leave the "Available" condition as false with the message "service has no endpoints" when a service's ingresses are ready before its endpoints.

imcsk8 commented:
Looks ok, just rebase and squash.

ironcladlou commented:
This is probably hopelessly out of date; going to close it and revisit later.

/close
